### ISSUE TITLE
Performance problems of lumina-fm

### ISSUE THREAD

**Issue by utrenkner** [trueos/lumina#358]
I have a problem with lumina-fm beeing so slow that I have to consider using another file manager.

The problem seems to be related to the number of files (and/or directories) in a directory. I have one particular "download" directory in which I keep a pile of files which I downloaded here and there, and for which I have no better place. Opening this dir in lumina-fm takes several seconds.

Even worse: Any file operation then takes much, much longer to execute. I measured the time to delete a single 400 kB PDF from the problematic directory: 50 seconds to show in lumina-fm, plus an additional ~40 seconds until the file manager was responding again to user interaction (selecting another file). Same behaviour for renaming.

The laptop is a four year old T530, with a i7 CPU and 8GB of RAM. Lumina was compiled from source (currently 149b0b98) on FreeBSD 11.0-RELEASE. Other software on the same laptop has no problem whatsoever with this directory (e.g. when opening it in evince to find a PDF, the content is presented instantly; same if I want to attach a file in Thunderbird from this directory - the content is shown without any noticeable delay).

There must be something terribly inefficient in the way lumina-fm handles this directory - possibly related to the ZFS snapshot feature(?). The directory currently has only 8 ZFS snapshots and no additional datasets or snapshots below it. Other directories are somewhat slow, but still usable. The problem really seems to be related to the size / number of files. The problem has existed for some time - possibly even since my first using of lumina (version 0.9). But I only started snapshotting later - so the problem may have become much bigger over time.

I also tested and this time it has nothing to do with the home-directory being a symlink. Opening /usr/home/uwe/downloads shows the same behaviour.

**Comment by utrenkner** [comment]
After my post, I just tested: Even after destroying all 8 snapshots of /usr/home/uwe/downloads does everything remain very slow in this directory. So it seems to be related to the size / number of files, rather than snapshots.

**Comment by utrenkner** [comment]
Did some more testing: I moved most of the files into sub-directories. And this seems to do the trick. The performance of lumina-fm in the downloads folder is much(!) better now. But still considerably below other tools. 

The problem clearly seems to be the **number** of files directly in this directory. Moving them to sub-directories speeds up lumina-fm. Furthermore, it seems the overall size (in GB) does **not** seem to matter. I had a few very large files in the download folder (e.g. a 4+ GB ISO file of the Win10 installation disk). But moving that one, too, did not seem to speed up lumina-fm any further.

Hope this helps identifying the problem.

**Comment by beanpole135** [comment]
It is probably related to the multi-threading of lumina-fm:
It run the item loader/parser in a separate thread for *every* item in the directory - so depending on the limit your system has for the number of concurrent system threads, it might be causing a general slowdown. 

I just committed a fix for this a moment ago. Please see if it makes any difference on your system. Basically, I just changed it so that a new thread is only spun up for every image file or *.desktop file (since those files need to actually be opened/read before the item becomes available).

**Comment by utrenkner** [comment]
Thanks. It does improve the behaviour slightly. But overall, performance of file operations are still delayed by up to 50 seconds.

In order to have a clear test case, I created a new directory /tmp/testdir with 500 empty files in it. Similar problem here: Renaming one file in lumina-fm takes ca. 15 seconds (until it shows the change) plus another ca. 10 seconds until the file manager becomes usable again. 

You may want to reproduce it for yourself - I used this little script to generate 500 empty files under /tmp/testdir. I used the suffix ".txt" so that lumina-fm would not have to test it as an image file.
```
#!/bin/sh
mkdir /tmp/testdir
i=1000
while [ $i -lt 1500 ]
do
	touch /tmp/testdir/$i.txt
	i=`expr $i + 1`
done;
```

**Comment by beanpole135** [comment]
Strange, on both my laptop and my desktop I use one of the Oxygen icon directories for testing the file manager (/usr/local/share/icons/oxygen/32x32/actions : ~650 *.png files), and in both cases it takes <2 seconds to load the directory (this should be the worst case scenario - needs to open/read every single file).

Which filesystem are you using on FreeBSD 11.0? I am using ZFS here, but I have not been testing under UFS at all.

**Comment by beanpole135** [comment]
Ok, I think I tracked down the issue and committed another fix for it.
It looks like Qt was not re-using previously loaded QIcons, so the generic mimetype icon was getting re-loaded into memory for *every* item in the directory. If I recall properly this was one of the special advantages for the QIcon loading mechanisms - so possibly a regression in Qt5? I added my own QIcon cache into the Browser backend for the mimetype icons, and that seems to significantly speed up the loading of the directory (I used your script and bumped it up to 1500 files  to test it out: <1 second or so).

**Comment by utrenkner** [comment]
Thank you. I built again from latest sources, but unfortunately do not see any improvement: With 1500 test files it takes ca. 20 seconds each to show the renamed file name and to become responsive again. 40 seconds for something that should normally work in less than a second is not good. On the shell the same renaming is executed instantly.

This is stock FreeBSD 11.0-RELEASE-p5 with GELI full-disk encryption using ZFS (as offered by the installer). The drive is a ca. 1 year old Samsung 840 PRO SSD with 512 GB.

I also tested on another internal SSD (Intel), which holds the previous FreeBSD 9.3 installation on UFS (with journaled soft-updates). This one is unencrypted. The result is comparable. So it does not have anything to do with either the filesystem or the encryption.

**Comment by utrenkner** [comment]
I just looked for a light-weight graphical file manager to compare with. I installed [x11-fm/pcmanfm](https://www.freshports.org/x11-fm/pcmanfm/) and this one does not have any noticeable delay when performing basic file operations such as renaming or deleting. Even the opening of the test directory seemed instantaneous - and thus quite different from lumina-fm which takes a few seconds.

Is there any output that I could log and send you? I am not a programmer and also have no experience with DTRACE - but wouldn't this be a perfect case for tracing lumina-fm in order to find out what is taking so long?

**Comment by q5sys** [comment]
You can try running truss ` truss -fed -o insight.txt lumina-fm /tmpdir`
Where /tmpdir is the dir that you're testing with.  Be warned, the file that truss is outputting to may end up being huge.  You'll prob need to compress it and host it somewhere.

**Comment by utrenkner** [comment]
You are right: The original log is 17 MB. I cut out 200,000 lines which practically repeated this line for 13 seconds:
```
0.344791916 close(17)                     ERR#9 'Bad file descriptor'
```
Apart from this, I noticed quite a few oxygen-icons which are not found... I cut after 100 seconds because the rest looked very uninteresting (basically the end is repeated over and over).

Here is the cut-down file: [fm-truss.txt](https://github.com/trueos/lumina/files/716546/fm-truss.txt)

**Comment by q5sys** [comment]
Ok I checked the truss output.  Couple things.
1) run it again, and look for where you get something like:
`10235: 99.761902630 pselect(0x19,0x7fffdf9facd0,0x0,0x0,0x7fffdf9fac30,0x0) = 0 (0x0)`
`10235: 99.762207093 kill(10254,SIGTERM)		 = 0 (0x0)`
When you run it again the pid that's getting killed will be different, take whatever pid shows up and let us know what process that is.

2) Have you installed the KDE oxygen icon set?

3) Next time... post the whole log somewhere (or email it to me jt @ ixsystems ) I want to see if I can figure out where that 'Bad File Descriptor' is coming from, but I'll need the whole log.

**Comment by utrenkner** [comment]
1. I know exactly what that process was - the command that I truss'ed: ``lumina-fm /tmp/testdir`` . I know this because I later could not start lumina-fm and was trying to kill that process - unsuccessfully. I had to reboot in order to get rid of that hanging process.

2. Yes, kde4-icons-oxygen-4.14.3 is installed from the official FreeBSD quarterly repo. But apparently lumina expects specific icons, which are not installed. Could that somehow be a mismatch because I compiled lumina from source, whereas everything else is from the FreeBSD repo?

3. I was looking for your e-mail address and could not find it... But for the bad file descriptor you have everything you need, I believe. I only cut ca. 200,000 (almost) identical lines relating to the bad file descriptor. I did not cut anything before it!

**Comment by q5sys** [comment]
1) You were unable to kill the lumina-fm process with a kill -9?
2) By default Lumina uses the KDE Oxygen icon set, and we pull that from the ports tree as well, so there should be no issue there.  Not sure why you're getting file not found errors.  I'll run a check on my system to see if I can figure anything out with regards to that.
3) Shouldnt have had to look for my address, all you had to do was remove the spaces and add the .com at the end. :) I didnt want to write it all out and it get caught by some spam robot searching for valid email addresses.
I assumed the bad file descr. was at the end of the file, once i went back to the beginning I saw it. I'm going to go over the truss output right after lunch and I may have some other ideas.

**Comment by utrenkner** [comment]
Just sent you a new truss output via e-mail (I had looked for your e-mail on your website _before_ you posted the ixsystems address here).

1. Yes, that is correct. The process survives any of my attempts to kill it - even with a kill -9!

2. I just looked up a few of those missing files, e.g. ``/usr/local/share/icons/oxygen/256x256/apps/go-first-view.png``. But that does not seem to exist in the port's [pkg-plist](https://svnweb.freebsd.org/ports/head/x11-themes/kde4-icons-oxygen/pkg-plist?revision=344893&view=markup). Nether does any SVG file, such as ``/usr/local/share/icons/oxygen/256x256/actions/go-first-view.svg``. So it is not surprising, that they are not found. But strange that Lumina somehow expects them there.

**Comment by q5sys** [comment]
I'm getting different issues with icons, but its not really an issue.  They are eventually found and the total time from their initial request to them being found is 0.008 seconds.  That's not anything we need to worry about.

The issue of the non-kill-able process is an oddity that's come up with the way Truss is trying to trace everything.  In real world use of lumina-fm, this isn't going to happen.

I've test multiple times today, and I'm not getting the same behavior that you are.  The longest delay I've gotten in launching lumina-fm was about 4 seconds.  And that was launching from nothing directly into the /tmpdir.  If I launched in to any other dir and then navigated to that directory the delay was under 2 seconds.

As for the 'Bad File Descriptor' errors.  I am getting those locally as well, but overall time spend on them is under 3 seconds.  So that doesn't account for the delay you're seeing.  I traced that and Ken and I stepped through the trace along side the lumina-fm source and we believe that it's something QT is doing during the process initiation.  We 'think' it might be related to the qt process watching, but aren't entirely sure.  But regardless, that wouldn't account to the delay you are seeing.

Since neither Ken nor I are unable to reproduce this on our end... we're not sure what other options we have to try to figure this out.  We'll continue to think on it.

**Comment by utrenkner** [comment]
Thank you very much for the effort. I just asked on bsdforen.de if others who use Lumina could test this, too. Let's see, if we can get useful feedback.

Thereby, I noticed that there exists a [Qt-Version of PCManFM](https://www.freshports.org/x11-fm/pcmanfm-qt/). Out of curiosity, I gave it a try: Again, no problems with opening the large test directory.

**Comment by utrenkner** [comment]
We got a [first confirmation of the problem on bsdforen.de](https://www.bsdforen.de/threads/lumina-fm-bitte-einmel-helfen-und-testen.33245/#post-288371). So my case is not the sole exception.

The system in question was here: 11.0-RELEASE, i7 2700K, 16GB RAM, ZFS on a HDD.

**Comment by q-pa** [comment]
Fresh install of TrueOS-2017-06-01-x64 STABLE on Lenovo Thinkpad T520, using integrated graphics with intel driver.

I, too, am experiencing performance problems. I do not have a directory on my local HDD with that many files but when I try to open a directory on my NAS share with directories AND files (couple of hundreds) Lumina-FM slows down (top shows WCPU values between 70 and 90%). (FYI: Directories with directories only as well as mixed directories with, let's say 130 files do not slow down Lumina-FM.)

Furthermore, I checked the system messages log file (see attached) and found the following lines:
...
Jun  6 20:54:12  kernel: swap_pager_getswapspace(16): failed
Jun  6 20:54:12  kernel: swap_pager_getswapspace(3): failed
Jun  6 20:54:12  kernel: swap_pager_getswapspace(4): failed
Jun  6 20:54:30  kernel: pid 13852 (lumina-fm), uid 1001, was killed: out of swap space

`gpart show` shows 4 GB of swap space
I did not try an alternative file manger yet.

**Comment by q-pa** [comment]
I did some more testing:
- using the oxygen icon set does not speed things up
- no difference between LAN and WIFI 
- opening directories with files that do not create thumbnails there is noticeable increase in speed
- depending on the number of files it can take minutes to open a directory
- I tried PCManFM: like @utrenkner mentioned, there is no delay at all! 
- 170607_console.txt shows the output of lumina-fm opening a directory of 409 mixed files
- 170607_messages.txt is the complete messages log of the last few days

[170607_console.txt](https://github.com/trueos/lumina/files/1058928/170607_console.txt)
[170607_messages.txt](https://github.com/trueos/lumina/files/1058930/170607_messages.txt)

FYI: I was using PCBSD 10.3 on that very same machine and did not have this problem and when I turned off the thumbnail view I noticed an additional increase in speed.

**Comment by utrenkner** [comment]
The performance issue seems to be relatively widespread. On the German bsdforen.de we recently had [another confirmation of this problem](https://www.bsdforen.de/threads/lumina-fm-bitte-einmel-helfen-und-testen.33245/#post-291165). 

Looking at the number of reports, I am really surprised that the developers apparently cannot lay their hands on any machine/installation, which is affected by this issue.

**Comment by q-pa** [comment]
This is strange, indeed. In terms of hardware at hand, I know for a fact that @kmoore134 actually had the exact same laptop I have before he got himself an X1 Carbon. Hopefully, he has that thing is still lying around for @beanpole135 to test on.

Furthermore, I am happy to provide more Information on my system, of course. @beanpole135, @q5sys:  Just tell me what you need.

**Comment by q5sys** [comment]
We've changed a few things this week that may have an impact on this. If you two can retest and report back we'd really appreciate it.

**Comment by q-pa** [comment]
Happy to! Many thanks! I assume these changes only affect the UNSTABLE repo. Can I safely switch to UNSTABLE, then update & test and afterwards go back to my last BE and switch back to STABLE?

**Comment by q5sys** [comment]
You'd need to download the source and compile it yourself, there are instructions in the main README in the repo root dir.  It's not hard. Download the repo, unzip, go into the directory, run qmake, then make && make install.

**Comment by q-pa** [comment]
Not hard for you, maybe. ;) OK, I have the repo. So far so good. 
Two questions: 1. Can I really install Lumina from within Lumina? 2. Should I revert back to Lumina 1.2.2 by installing that from the AppCafe when I am done with testing to be back on clean STABLE or will the next STABLE updates just bring me back there? Easiest would be going back to the backup BE, I guess.

**Comment by DevHPX** [comment]
@q5sys Gostei

**Comment by q-pa** [comment]
Do the last STABLE updates include all changes from 9 days ago? I definitely like the directory tree and the ability to choose whether to load thumbnails or not but the original problem of this post is not resolved. 
Attached you find the log of starting lumina-fm from a terminal. The messages have changed a bit (10 days ago: https://github.com/trueos/lumina/files/1058928/170607_console.txt). But the rule still is: the more files in the directory, the longer it takes to open it. 
(I did not test a week ago because I need to  be sure to be able to revert back to STABLE).

**Comment by q-pa** [comment]
===================
General information
===================

boot environment now (N) … 12.0-CURRENT-up-20170704_202439 NR 2017-07-04
       after restart (R) … 12.0-CURRENT-up-20170704_202439 NR 2017-07-04
boot loader …………………………………… BSD
            type ……………………… EFI
CPU ………………………………………………………… Intel(R) Core(TM) i7-2670QM CPU @ 2.20GHz
    number of cores ……………… 8
host ……………………………………………………… trueos-L
memory ………………………………………………… 16384 MB available, 14092 MB free
OS git branch ……………………………………………………………………………………… trueos-stable
OS git revision ………………………………………………………………………………… ac2f0aa3b
OS kernel build time ………………………………………………………… Wed 2017 Jun 21 01:09:23 UTC
OS kernel identity …………………………………………… (uname -i) GENERIC
OS platform (architecture) ……………………… (uname -m) amd64
OS release level ………………………………………………… (uname -r) 12.0-CURRENT
OS version and patch level …… (freebsd-version) 12.0-CURRENT
TrueOS package set ………………… STABLE
TrueOS version …………………………… TrueOS-Desktop-201706210948
uptime ………………………………………………… 16 mins
user  …………………………………………………… delipu

More (TrueOS Desktop):

    desktop environment …… Lumina
    sound card driver ………… pcm0: <Conexant CX20590 (Analog 2.0+HP/2.0)> (play/rec) default
    wireless driver ……………… iwn0
    X11 drivers ………………………… intel_drv.so

Since the last updates there are no error messages any more. Opening the directory is much quicker now but it still takes much longer than under PCBSD 10.3 (1-2 seconds).

TrueOS:
Before update: several minutes
After last update: 18 seconds (PCManFM: 9 seconds)

**Comment by q-pa** [comment]
Hm, cannot edit comment with qupzilla, so more information in this comment.

The test from my last comment described accessing a samba share on my NAS containing 7 directories and 725 files (mounted via sudo mount_smbfs -L de_DE.UTF-8 -N -Ixxx.xxx.xxx.xxx  //user@NAS/directory /mnt/NAS/directory).

Opening /usr/bin  takes about 4 seconds.
Opening /usr/sbin takes about 3 seconds.

On PCBSD 10.3 it was more or less instanteneous.

**Comment by q5sys** [comment]
I've talked to Ken about changing the underlying model we are using for querying the file system, which hopefully will be the final step needed to resolve this.  It's a bit of an undertaking though, so it might take a bit.  I've been poking at it here and there, but it'll take me getting the time to sit down and focus completely on it.  Thanks for the continued information. It helps to know we are heading in the right direction with our updates.

**Comment by utrenkner** [comment]
Sorry for the late reply - I did not find the time for tests so far. I now compiled the latest sources (c7a44d46) on my FreeBSD 11.0-RELEASE laptop and tested on a directory with 1500 files created with the little script mentioned above.

The lumina file-manager feels like opening the directory much quicker than before. But file operations are still painfully slow. I checked with my wrist watch: Renaming a file in lumina-fm took ca. 30 seconds (showing the new file name) and pretty much the same time, until lumina-fm was responsive again (selecting another file). 1 minute for such a seemingly simple operation is still unbearable for everyday work, unfortunately. So, I will keep the mate desktop as my DE of choice, for the time being.

Just checked: With the PCManFM file manager, the same operation is so quick, that I cannot even measure it with my wrist watch.

**Comment by q5sys** [comment]
I'm pretty sure this is tied to the fact that PCManFM is using the QFileSystemModel for the back end.  I'm working on integrating that into Insight, but it's going to take a bit of time.  Hopefully that will resolve all of these performance problems.  I dont see anything else in the code that could be giving us this type of performance degradation.  The other question is... how is it that we are unable to reproduce this?  I've tried Lumina across several dozen systems personally, and I've never experienced a 30 second delay.   What type of hard drive do you have?  Is it in an idle state before you try the file operation?

**Comment by q-pa** [comment]
Thank you very much for further investigating this! It is an off-the-shelf laptop with a spinning disk hard drive.

dmesg:
...
ada0 at ata0 bus 0 scbus0 target 0 lun 0
ada0: <TOSHIBA MK5061GSY MC102E> ATA8-ACS SATA 2.x device
ada0: Serial Number 73CZT6CUT
ada0: 300.000MB/s transfers (SATA 2.x, UDMA5, PIO 8192bytes)
ada0: 476940MB (976773168 512 byte sectors)
...

It was never in idle state before I tested to access a file system.

**Comment by utrenkner** [comment]
My drive is a Samsung 840 PRO SSD, running full-disk encrypted ZFS. I had tested an earlier version of lumina-fm also on a directory on an older Intel SSD running UFS, unencrypted. Same result.

**Comment by q5sys** [comment]
I just did a fresh test of this on a brand new fresh install using the 14 July ISO using your script above.  This was an extremely low powered system (Intel Nuc  Celeron N2830 with 2gb of RAM, and an old Transcend 32gb ssd. I went with a low powered system specifically so I could try to make any performance impact greater.

If lumina-fm was open and I navigated to that directly, it took ~ 1 second to load the entire directory and show the files.  If I launched Lumina-fm from the command line, it took ~1.5 seconds to load lumina-fm and the entire directory listing.  (Time not shown at shell until after I closed the file manager)

[q5sys@trueos-nuc] /tmp% time lumina-fm /tmp/testdir/
Loading Locale: "lumina-fm" "en" "UTF-8"
 - Could not load Locale: ""
 - Created new single-instance lock
Loaded QApplication
Externally-mounted devices: ()
1.537u 0.768s 0:02.70 84.8%     1404+229k 10812+3io 0pf+0w
[q5sys@trueos-nuc] /tmp%

**Comment by utrenkner** [comment]
What OS where you testing? Vanilla FreeBSD or TrueOS? I know it *shouldn't* make a difference, but maybe TrueOS comes with certain sysctls or other settings, that help avoid this problem.

**Comment by q5sys** [comment]
I'm testing this under the TrueOS iso image from 14 July (see link below), what OS are you using?

http://download.trueos.org/unstable/amd64/TrueOS-2017-07-14-x64-DVD.iso

**Comment by Mrt134** [comment]
Since this is stalled for 1+ month and we can't seem to reproduce the issue here, I'm going to close as "unable to reproduce". Leave an update or comment if I need to reopen. Thanks!

**Comment by q-pa** [comment]
Since I am on UNSTABLE, performance is better with the local file system (i.e. as quick as PCManFM-qt) but still quite faulty when performing operations on my remote shares (smbfs).

1. : Opening home directory (< 1 sec):
```
delipu@trueos-L] ~% lumina-fm
lthemeengine: using lthemeengine plugin
Loading Locale: "lumina-fm" "de_DE" "UTF-8"
 - Created new single-instance lock
Loaded QApplication
lthemeengine: D-Bus global menu: no
Externally-mounted devices: ("UNKNOWN::::fusefs::::/usr/home/delipu/.gvfs")
libpng warning: iCCP: known incorrect sRGB profile
```

2. Opening /etc (about 1 sec): 
```
[delipu@trueos-L] ~% lumina-fm
lthemeengine: using lthemeengine plugin
Loading Locale: "lumina-fm" "de_DE" "UTF-8"
 - Created new single-instance lock
Loaded QApplication
lthemeengine: D-Bus global menu: no
Externally-mounted devices: ("UNKNOWN::::fusefs::::/usr/home/delipu/.gvfs")
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
QKqueueFileSystemWatcherEngine::addPaths: open: Permission denied
```

3. Opening a directory on the smbfs share (this took about 30 sec for a directory containing 726 files; opening directories with just directories is more or less instantaneous):
```
[delipu@trueos-L] ~% lumina-fm
lthemeengine: using lthemeengine plugin
Loading Locale: "lumina-fm" "de_DE" "UTF-8"
 - Created new single-instance lock
Loaded QApplication
lthemeengine: D-Bus global menu: no
Externally-mounted devices: ("UNKNOWN::::fusefs::::/usr/home/delipu/.gvfs")
```

File operations on smbfs share tend to crash from time to time. Cutting file operations never work, leaving temporary files in the directory. I mount the shares exactly like I did under PCBSD 10.3 where I had no problems at all.

**Comment by q-pa** [comment]
I will update this issue after the next UNSTABLE update.

**Comment by Mrt134** [comment]
Thank you!

**Comment by q-pa** [comment]
The update to Lumina 1.3.3 did not solve the problem. PCManFM seems to work better, though since either the latest UNSTABLE or the Lumina update. I fear I have to work with that until this issue is solved... If you need anything like e.g. logs please feel free to ask. I will try to provide whatever information you need.

**Comment by utrenkner** [comment]
I can confirm @q-pa . On my T530 with vanilla FreeBSD 11.1-RELEASE-p1, the performance may have improved slightly... but it still takes ca. 20s each for a rename to show and for lumina-fm to become responsive again.

**Comment by Mrt134** [comment]
Thanks for the update - we had a setback last week that slowed us down, but this is still on the "to-do" list. :)

**Comment by utrenkner** [comment]
Would be great - because this problem seems much more widespread than previously thought. Other users just reported their tests of lumina-fm on the German bsdforen.de  (using my [little script](https://github.com/trueos/lumina/issues/358#issuecomment-273060585) to create 1500 empty files). 

One of them concluded: "unusable".

Another one tested lumina-fm on FreeBSD 11.1-RELEASE or -STABLE on three different laptops:
- [Asus Eee PC 1001PX](https://www.bsdforen.de/threads/lumina-fm-bitte-einmel-helfen-und-testen.33245/#post-295177) -> deleting or renaming took 2 minutes, each, during which WCPU was at 76-77%
- [Toshiba Satellite A300](https://www.bsdforen.de/threads/lumina-fm-bitte-einmel-helfen-und-testen.33245/#post-295176) -> renaming took 1 minute, during which WCPU was at 68–69%
- [Lenovo ThinkCentre M58p](https://www.bsdforen.de/threads/lumina-fm-bitte-einmel-helfen-und-testen.33245/#post-295178) -> renaming took ca. 1/2 minute, during which WCPU was at ca. 70%

This really is a showstopper for people who contemplate using Lumina on their FreeBSD systems.

**Comment by beanpole135** [comment]
I just tested Insight using your script above (500 empty text files in /tmp/[testdir]) on my own Asus Eee PC, and it only took 3 seconds to load the entire directory.

Hardware: (Asus Eee PC )
 * CPU: Intel(R) Atom(TM) CPU N450 @ 1.66GHz (2 physical cores, 1662.53-MHz K8-class CPU)
 * Memory: 2048 MB available, 1342 MB free

Software:
 * TrueOS UNSTABLE (fresh install with no modifications on entire hard drive)
 * OS Version: FreeBSD 12.0-CURRENT #82a2a3070d
 * Filesystem: OpenZFS (entire system)
 * Lumina version: 1.3.3

I will try the same test on a stock FreeBSD system here in a little bit (UFS and/or ZFS) to see if that narrows down where the problems are coming from.

**Comment by beanpole135** [comment]
Ok, here are the results of the FreeBSD 11.1/ZFS installation on the same Asus Eee PC machine:

Result:
* Time to load directory: 3.5 seconds

System setup instructions:
1. Install FreeBSD 11.1-Release (with auto-ZFS layout)
2. login as root and run `pkg install xorg lumina` to install everything
3. as root, run `dbus-uuidgen --ensure` to create the machine-id file/flag that Qt needs
3. login as user and run `start-lumina-desktop` to launch the desktop
4. create/run the file-generation script that is referenced above
5. launch Insight in /tmp, then change to the test directory for timing test

Differences with the TrueOS test:
* Lumina Version 1.3.0
* FreeBSD version 11.1 instead of 12.0-CURRENT

I will be re-doing the system/test with a UFS setup here in just a moment.

**Comment by beanpole135** [comment]
Results of the same test for a fresh FreeBSD 11.1/UFS install:

Result:
* Time to load directory: 2.5 seconds

System setup instructions:
1. Install FreeBSD 11.1-Release (with auto-UFS layout)
2. login as root and run pkg install xorg lumina to install everything
3. as root, run dbus-uuidgen --ensure to create the machine-id file/flag that Qt needs
4. login as user and run start-lumina-desktop to launch the desktop
5. create/run the file-generation script that is referenced above
6. launch Insight in /tmp, then change to the test directory for timing test

Differences with the TrueOS test:
* Lumina Version 1.3.0
* FreeBSD version 11.1 instead of 12.0-CURRENT
* UFS filesystem rather than ZFS

**Comment by q-pa** [comment]
boot environment now (N) … 12.0-CURRENT-up-20170926_204316 NR 2017-09-26
       after restart (R) … 12.0-CURRENT-up-20170926_204316 NR 2017-09-26
boot loader …………………………………… BSD
            type ……………………… EFI
CPU ………………………………………………………… Intel(R) Core(TM) i7-2670QM CPU @ 2.20GHz
    number of cores ……………… 8
host ……………………………………………………… trueos-L
memory ………………………………………………… 16384 MB available, 13080 MB free
OS git branch ……………………………………………………………………………………… drm-next
OS git revision ………………………………………………………………………………… 2a2a3070d
OS kernel build time ………………………………………………………… Fri 2017 Sep 22 13:48:41 UTC
OS kernel identity …………………………………………… (uname -i) GENERIC
OS platform (architecture) ……………………… (uname -m) amd64
OS release level ………………………………………………… (uname -r) 12.0-CURRENT
OS version and patch level …… (freebsd-version) 12.0-CURRENT
TrueOS package set ………………… UNSTABLE
TrueOS version …………………………… TrueOS-Desktop-201707111911
uptime ………………………………………………… 11 mins
user  …………………………………………………… delipu

More (TrueOS Desktop):

    desktop environment …… Lumina
    sound card driver ………… pcm0: <Conexant CX20590 (Analog 2.0+HP/2.0)> (play/rec) default
    wireless driver ……………… iwn0
    X11 drivers ………………………… intel_drv.so

----------------------------------------------------------------------------------------------------------
I just tested it again with the abovementioned script: 
Lumina-FM: about 2s
PCManFM: about 1s

When accessing remote directories on my NAS the difference is huge though:
Lumina-FM: several minutes (sometimes crashing)
PCManFM: a few seconds

**Comment by utrenkner** [comment]
@beanpole135 Thank you very much for testing on different OS versions!!!

Just so that we are looking at the same thing: The problem I and others have is not with the initial opening/viewing of a directory in Insight (lumina-fm). The problem becomes most apparent when making a file operation.

My test case looks like this (lumina is installed from official FreeBSD quarterly repository on FreeBSD 11.1-RELEASE).

1. login to FreeBSD as user
2. ``start-lumina-desktop``
3. prepare test directory with script above
4. open Insight and go to that test directory

[Now comes the actual test]

5. right-click a file and rename it (e.g. 1001.txt into 1001a.txt)
6. click on another file (immediately after committing the name change; or after the name change becomes visible in Insight)

Result: After step 5 it takes Insight a long time (20-30 seconds) to actually show the change of the name. During this time, Insight does not react to any kind of input.

After step 6 it takes another 20-30 seconds until the newly clicked file becomes selected and Insight reacts to user input.

So, step 4 clearly is not a huge problem (even though I have the impression other file managers are much faster, loading even the large test directory almost instantly). However, it is the renaming (or other file operations) which are really crippling the performance.

Interestingly, if I understood it correctly, one user on bsdforen.de did the test a little differently, but with the same result: Hi changed the filename in another file manager and waited for the change to become visible in Insight... I have not tested this. It would mean that the problem lies not in actually changing the file but in showing the change in Insight.

And of course, the high WCPU value during the change in Insight should be of interest. What is the software doing, then?

**Comment by utrenkner** [comment]
And one more test: On a T430 running the latest HardenedBSD 11-STABLE and lumina-1.3.0,3 from the HBSD official repo. I did not take the exact time - but it generally showed the same delays after step 5 and after step 6.

All the latest tests were on ZFS, but an earlier test on UFS showed no difference.

**Comment by q-pa** [comment]
I can confirm the results of the renaming test.

**Comment by beanpole135** [comment]
Since I still have that Asus Eee PC setup with the FreeBSD 11.1/UFS, I just booted it back up and tried the modification tests you mentioned:

Results:
* Open directory: ~2.5 seconds
* Rename single file: ("1001.txt" -> "1001a.txt") : 11 seconds
* Remove single file: (1001a.txt) : 14 seconds
* Create single File: (1001a.txt): 14-15 seconds

That is with the 1.3.0 version of Lumina as installed by FreeBSD packages.
Since I see the delay you are talking about, let me try out some of the alternative options (update to Lumina 1.3.3, look into any dir-update operations that can be streamlined, etc..) and see if I can get that time down for you.

**Comment by beanpole135** [comment]
Fixed!
https://github.com/trueos/lumina/commit/860d2ca1863f3838b5575c6efc2151a7270b0d79

For verification, I rebuilt the file manager with this change on the FreeBSD 11.0/UFS system, and the modification "update" times dropped to <1 second.

**Comment by utrenkner** [comment]
Great! For what concerns the file operations. They are now so fast, I could not clock them at all!!!

However, the problem has now shifted to the initial opening of the folder. While that was never a huge problem, my test folder with 1500 files now takes 15 seconds to open. So it seems you solved one problem but introduced another one.

Unfortunately, until both cases are solved, Lumina cannot become my daily desktop... Do you want me to open a new issue for this, or should we continue here?

**Comment by q-pa** [comment]
Is the fix part of yesterday's UNSTABLE update?

**Comment by beanpole135** [comment]
Yes - these fixes are included in the UNSTABLE updates to TrueOS that got built/pushed a couple days ago.

**Comment by q-pa** [comment]
I am sorry to tell you that it got worse with the latest update.
Now Lumina-FM just crashes when I try to access a directory on my NAS. When started from the command line, this is the output:
`~% lumina-fm
lthemeengine: using lthemeengine plugin
Loading Locale: "lumina-fm" "de_DE" "UTF-8"
 - Created new single-instance lock
Loaded QApplication
lthemeengine: D-Bus global menu: no
Externally-mounted devices: ()
libpng warning: iCCP: known incorrect sRGB profile
Bus error
`

**Comment by q-pa** [comment]
I tried to insert code but it did not really work...

~% lumina-fm
lthemeengine: using lthemeengine plugin
Loading Locale: "lumina-fm" "de_DE" "UTF-8"
 - Created new single-instance lock
Loaded QApplication
lthemeengine: D-Bus global menu: no
Externally-mounted devices: ()
libpng warning: iCCP: known incorrect sRGB profile
Bus error

**Comment by q-pa** [comment]
Sometimes also crashes in local  directories.

**Comment by q-pa** [comment]
FYI: dbus is running.

**Comment by q-pa** [comment]
Just updated to the latest UNSTABLE release. Lumina-FM is running much smoother and opening the NAS share I tested with before now takes about 15s. (PCManFM is still twice as fast). 

No crashes so far.

**Comment by utrenkner** [comment]
Happy New Year to all Lumina developers!

The new year brought a new quarterly repo on FreeBSD and thus lumina-1.4.1,3. I installed it, and was super happy to see that the performance problems of lumina-fm are gone! Both, on FreeBSD 11.1-RELEASE and on HardenedBSD 11-STABLE. The file operations which lasted for several tens of seconds now take only a blink of an eye - just like in other file managers.

Great work. Many thanks staying with me on this long-timed issue and for finally working it out. I now switch back to trying Lumina as my daily DE!

**Comment by q-pa** [comment]
Just for your information: with Trident this issue is completely resolved for me.
